### ISSUE TITLE
It is unclear how to display a line/column number on a NamedSource

### ISSUE THREAD

**Issue by brassy-endomorph** [zkat/miette#208]
In the output of the following code, I would expect the line/column numbers following the file name to be equal to the location of the error, but this isn't the case and from looking at the examples, I can't find a way to do this. I would be nice if this was documented somehow as the screenshots in the read me seem to have this error as well.

# Minimal example to reproduce

```bash
$ rustc --version
rustc 1.64.0 (a55dd71d5 2022-09-19)
$ cargo --version
cargo 1.64.0 (387270bc7 2022-09-16)
```

### `Cargo.toml`

```toml
[package]
name = "miette-test"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
miette = { version = "5.3.0", features = ["fancy"] }
thiserror = "1.0.37"
```
### `main.rs`

```rust
use miette::{Diagnostic, SourceSpan, NamedSource, Report};

pub type Result<T> = ::std::result::Result<T, Error>;

#[derive(Diagnostic, thiserror::Error, Debug, PartialEq)]
#[error("Some error: {kind}")]
pub struct Error {
    pub kind: ErrorKind,
    #[label = "error"]
    pub source_span: SourceSpan,
}

#[derive(Diagnostic, thiserror::Error, Debug, PartialEq, Clone)]
pub enum ErrorKind {
    #[error("Something bad: {}", .0)]
    SomeKind(String),
}


fn main() -> miette::Result<()> {
    let program = "this is\na BAD program\n";
    Err(Error {
        kind: ErrorKind::SomeKind("oh no".to_owned()),
        source_span: (program.find("BAD").unwrap(), 3).into(),
    })  
    .map_err(Report::new)
    .map_err(|err| err.with_source_code(NamedSource::new("my_program.xyz", program)))
}
```

### Actual Terminal Output

```
Error: 
  × Some error: Something bad: oh no
   ╭─[my_program.xyz:1:1]
 1 │ this is
 2 │ a BAD program
   ·   ─┬─
   ·    ╰── error
   ╰────
```

### Expected Terminal Output

```
Error: 
  × Some error: Something bad: oh no
   ╭─[my_program.xyz:2:3]    <---- line 2 column 3
 1 │ this is
 2 │ a BAD program
   ·   ─┬─
   ·    ╰── error
   ╰────
```

**Comment by CyriacBr** [comment]
This seems to be intended. See this https://github.com/zkat/miette/issues/91#issuecomment-950165566

**Comment by zkat** [comment]
Yeah. I'd say I would take a PR for this, but the fact that miette allows multiple highlights in a snippet makes this much trickier. I still think what miette does right now is the "correct" thing, even if it's a bit unexpected sometimes.

I'll leave this open in case folks want to discuss other alternatives, though.

**Comment by brassy-endomorph** [comment]
I think that using the behavior in `rustc` would make sense.

E.g., 

```
error[E0308]: mismatched types
 --> src/main.rs:6:18
  |
6 |     takes_string(123);
  |     ------------ ^^^ expected `&str`, found integer
  |     |
  |     arguments to this function are incorrect
  |
note: function defined here
 --> src/main.rs:1:4
  |
1 | fn takes_string(s: &str) {
  |    ^^^^^^^^^^^^ -------
```

It uses the line/col where the error originates even if there are multiple snippets. Like it was mentioned in the other PR, it may be desirable to disable this via config.

In the interpreter example (`3 + "5"`), while there are three snippets/labels (int, operator, str), I would consider the second to the the span where the error originated, but either the `3` or the `+` being the span displayed in the formatter would make sense and be enough to let me more quickly jump to that error.

Understandably there's a slight explosion of permutations on how to format this all, plus the complexity of what the default should be. I suppose what I would like is:

- a way to mark a snippet as primary
- display line/col of the primary
- if only one snippet (regardless of being marked primary or not) is present, display that
- if ambiguous, display nothing

That might also be confusing and lead to unexpected behavior where someone would ask "why isn't the line/col showing in this snippet" both as an implementer and an end user.

Or I guess even just displaying the line/col per snippet?

I think simply disabling it might be annoying because in a parser, one might have 1 line with 10,000 columns and really want to know the column where the parser erred.

**Comment by zkat** [comment]
the last time we tackled this, the hangup was that miette doesn't _have_ the concept of a "primary" span. Should it just be the first span, with other spans being treated as "context" spans?
